imap_processing/hit/l1a/hit_l1a.py

```diff
@@ -40,26 +42,131 @@ def hit_l1a(packet_file: str, data_version: str) -> list[xr.Dataset]:
+
+    # Calculate mod 10 values
+    hdr_min_count_mod_10 = sci_dataset.hdr_minute_cnt.values % 10
+
+    # Reference mod 10 mapping to initialize data structure for species and
+    # energy ranges and add 8x15 arrays with fill values for each science frame.
+    num_frames = len(hdr_min_count_mod_10)
+    # TODO: add more specific dtype for rates (ex. int16) once this is defined by HIT
+    data_by_species_and_energy_range = {
+        key: {**value, "rates": np.full((num_frames, 8, 15), fill_value=-1, dtype=int)}
+        for key, value in MOD_10_MAPPING.items()
+    }
+
+    # Update rates for science frames where data is available
+    for i, mod_10 in enumerate(hdr_min_count_mod_10):
+        data_by_species_and_energy_range[mod_10]["rates"][i] = sci_dataset[
+            "sectorates"
+        ].values[i]
+
+    # H has 3 energy ranges, 4He, CNO, NeMgSi have 2, and Fe has 1.
+    # Aggregate sector rates and energy min/max values for each species.
+    # First, initialize dictionaries to store rates and min/max energy values by species
+    data_by_species: dict = {
+        value["species"]: {"rates": [], "energy_min": [], "energy_max": []}
+        for value in data_by_species_and_energy_range.values()
+    }
+
+    for value in data_by_species_and_energy_range.values():
+        species = value["species"]
+        data_by_species[species]["rates"].append(value["rates"])
+        data_by_species[species]["energy_min"].append(value["energy_min"])
+        data_by_species[species]["energy_max"].append(value["energy_max"])
+
+    # Add sector rates by species to the dataset
+    for species_type, data in data_by_species.items():
+        # Rates data has shape: energy_index, epoch, declination, azimuth
+        # Convert rates to numpy array and transpose axes to get
+        # shape: epoch, energy_index, declination, azimuth
+        rates_data = np.transpose(np.array(data["rates"]), axes=(1, 0, 2, 3))
+
+        species = species_type.lower()
+        sci_dataset[f"{species}_counts_sectored"] = xr.DataArray(
+            data=rates_data,
+            dims=["epoch", f"{species}_energy_index", "declination", "azimuth"],
+            name=f"{species}_counts_sectored",
+        )
+        sci_dataset[f"{species}_energy_min"] = xr.DataArray(
+            data=np.array(data["energy_min"], dtype=np.int8),
+            dims=[f"{species}_energy_index"],
+            name=f"{species}_energy_min",
+        )
+        sci_dataset[f"{species}_energy_max"] = xr.DataArray(
+            data=np.array(data["energy_max"], dtype=np.int8),
+            dims=[f"{species}_energy_index"],
+            name=f"{species}_energy_max",
+        )
+        # add energy index coordinate to the dataset
+        sci_dataset.coords[f"{species}_energy_index"] = xr.DataArray(
+            np.arange(sci_dataset.sizes[f"{species}_energy_index"], dtype=np.int8),
+            dims=[f"{species}_energy_index"],
+            name=f"{species}_energy_index",
+        )
```

Comment:
Was there reason for not returning?

Reply:
This function updates the dataset in-place so a return isn't necessary